### PR TITLE
Create and delete endpoints

### LINKED ISSUE #3335: [CAD/Project File Review] - Create and Delete endpoints for Review Requests
### Description

We need endpoints to create and delete review requests on parts. There should be endpoints in the parts router at /reviewRequest/:partId/create and /reviewRequest/:partId/delete that create and delete review requests. Review requests should be able to be created by any member, but should only be able to be deleted by the review request requester (the creator), the person whose review is being requested, or leadership.

### Acceptance Criteria

There exists endpoints at /reviewRequest/:partId/create and /reviewRequest/:partId/delete with the above described permissions. There are tests for functionality of both successful and unsuccessful calls, and there are postman tests.

### Proposed Solution

Follow NER conventions and look at existing endpoints

### Mocks

_No response_

### PR DESCRIPTION
**Pull request by nateci**
## Changes

- Created POST /reviewRequest/:partId/create endpoint to allow any member to create a review request on a part.
- Created POST /reviewRequest/:partId/delete endpoint to allow the requester, reviewer, or leadership to delete a review request.
- Implemented permissions to ensure only the requester, reviewer, or leadership can delete review requests.
## Notes

- Needed to add deletedDate to schema and migration.
- did not use transformer on create/delete services, since it would require me editing the transformer, therefore editing the interface which would get overwritten everytime prisma:reset is run, so i deemed just not calling the transformer was the best move. (however this means that the postman screenshots are looooooong).

## Test Cases

- case A: Test creating a review request (success scenario).
- edge Case: Attempt to create a review request when the part does not exist (should return error).
- case B: Test deleting a review request by the requester (success scenario).
- case C: Test deleting a review request by the reviewer (success scenario).
- case D: Test deleting a review request by a leadership user (success scenario).
- case E: Test that a non-involved user cannot delete the review request (should return error).
- edge Case: Attempt to delete a non-existent review request (should return error).

## Screenshots

Screenshot 2025-03-27 142526
Screenshot 2025-03-27 142531
Screenshot 2025-03-27 142540
Screenshot 2025-03-27 142545

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [x] All commits are tagged with the ticket number
- [x] No linting errors / newline at end of file warnings
- [x] All code follows repository-configured prettier formatting
- [x] No merge conflicts
- [x] All checks passing
- [x] Screenshots of UI changes (see Screenshots section)
- [x] Remove any non-applicable sections of this template
- [x] Assign the PR to yourself
- [x] No `yarn.lock` changes (unless dependencies have changed)
- [x] Request reviewers & ping on Slack
- [x] PR is linked to the ticket (fill in the closes line below)

Closes #3335

### CODE REVIEW

File: src/backend/src/controllers/part-review.controllers.ts [hunk 1]
```diff
@@ -159,4 +159,38 @@ export default class PartReviewController {
...
+
+      res.status(200).json(request);
+    } catch (error) {
+      next(error);
+    }
+  }
+
+  static async deletePartReviewRequest(req: Request, res: Response, next: NextFunction) {
+    try {
+      const { partId } = req.params;
```
Comment (chpy04) [inline src/backend/src/controllers/part-review.controllers.ts]: This should take in a reviewRequestId, and delete that specific review request

File: src/backend/src/routes/parts.routes.ts [hunk 2]
```diff
@@ -66,4 +66,13 @@ partsRouter.post(
...
 partsRouter.post('/common-mistake/:commonMistakeId/delete', PartReviewController.deleteCommonMistake);
 
+partsRouter.post(
+  '/reviewRequest/:partId/create',
+  nonEmptyString(body('reviewerId')),
+  validateInputs,
+  PartReviewController.createPartReviewRequest
+);
+
+partsRouter.post('/reviewRequest/:partId/delete', PartReviewController.deletePartReviewRequest);
```
Comment (chpy04) [inline src/backend/src/routes/parts.routes.ts]: Same here, should be reviewRequestId

File: src/backend/src/services/part-review.services.ts [hunk 3]
```diff
@@ -433,4 +434,81 @@ export default class PartReviewService {
...
+   */
+  static async createPartReviewRequest(partId: string, requester: User, reviewerId: string, organizationId: string) {
+    const part = await prisma.part.findUnique({
+      where: { partId }
+    });
+
+    if (!part) {
+      throw new NotFoundException('Part', partId);
+    }
+
```
Comment (chpy04) [inline src/backend/src/services/part-review.services.ts:683]: Should also check if the part has been deleted (if the dateDeleted is not null)

File: src/backend/src/services/part-review.services.ts [hunk 4]
```diff
@@ -433,4 +434,81 @@ export default class PartReviewService {
...
+    const hasAccess = await userHasPermission(requester.userId, organizationId, isAdmin);
```
Comment (chpy04) [inline src/backend/src/services/part-review.services.ts]: Can we just make this any member? There isn't really any issue with people making a request for a review so we shouldn't be limiting it

File: src/backend/src/services/part-review.services.ts [hunk 5]
```diff
@@ -433,4 +434,81 @@ export default class PartReviewService {
...
+          connect: { userId: requester.userId }
+        },
+        reviewerRequested: {
+          connect: { userId: reviewerId }
+        }
+      },
+      ...partReviewRequestQueryArgs(organizationId)
+    });
+
+    return createdRequest;
```
Comment (chpy04) [inline src/backend/src/services/part-review.services.ts]: This should be passed into a transformer

File: src/backend/src/services/part-review.services.ts [hunk 6]
```diff
@@ -433,4 +434,81 @@ export default class PartReviewService {
...
+
+  /**
+   * soft deletes an existing part review request if the requester, reviewer, or an admin initiates the request
+   * @param partId - the ID of the part whose review request should be deleted
+   * @param user - the user attempting to delete the review request
+   * @param organizationId - the organization ID to validate permissions
+   * @returns the soft-deleted and transformed PartReviewRequest
+   */
+  static async deletePartReviewRequest(partId: string, user: User, organizationId: string) {
+    const reviewRequest = await prisma.partReviewRequest.findFirst({
```
Comment (chpy04) [inline src/backend/src/services/part-review.services.ts]: This is why we need to take in a reviewRequestId. This code would just delete whatever the first review request in the part is, which is not what we want to do

File: src/backend/src/services/part-review.services.ts [hunk 7]
```diff
@@ -433,4 +434,81 @@ export default class PartReviewService {
...
+      where: { partId }
+    });
+
+    if (!reviewRequest) {
+      throw new NotFoundException('Review request', partId);
+    }
+
```
Comment (chpy04) [inline src/backend/src/services/part-review.services.ts:732]: Same thing here, check if this review request has already been deleted

File: src/backend/src/services/part-review.services.ts [hunk 8]
```diff
@@ -433,4 +434,81 @@ export default class PartReviewService {
...
+    const isRequester = reviewRequest.requesterId === user.userId;
+    const isReviewer = reviewRequest.reviewerId === user.userId;
+    const isLeader = await userHasPermission(user.userId, organizationId, isAdmin);
```
Comment (chpy04) [inline src/backend/src/services/part-review.services.ts]: can we make this any leadersip instead of only admins?

File: src/backend/src/services/part-review.services.ts [hunk 9]
```diff
@@ -433,4 +434,81 @@ export default class PartReviewService {
...
+    const softDeletedRequest = await prisma.partReviewRequest.update({
+      where: {
+        partReviewRequestId: reviewRequest.partReviewRequestId
+      },
+      data: {
+        dateDeleted: new Date()
+      },
+      ...partReviewRequestQueryArgs(organizationId)
+    });
+    return softDeletedRequest;
```
Comment (chpy04) [inline src/backend/src/services/part-review.services.ts]: Same thing here, this should be in a transformer

**Review by chpy04 — CHANGES_REQUESTED** [review]
Looks really good, just needs a few tweaks

File: src/backend/src/services/part-review.services.ts (hunk 3 above)
Comment (chpy04) [reply src/backend/src/services/part-review.services.ts:683]: We actually have a specific DeletedException on top of the NotFoundException, so you should use that one here

File: src/backend/src/services/part-review.services.ts [hunk 10]
```diff
@@ -433,4 +434,86 @@ export default class PartReviewService {
...
+    const part = await prisma.part.findUnique({
+      where: { partId }
+    });
+
+    if (!part || part.dateDeleted) {
+      throw new NotFoundException('Part', partId);
+    }
+
+    const hasAccess =
+      (await userHasPermission(requester.userId, organizationId, isMember)) ||
```
Comment (chpy04) [inline src/backend/src/services/part-review.services.ts]: Permissions are hierarchical: admins and leadership are also members, so you can get rid of the admin and leadership tests

File: src/backend/src/services/part-review.services.ts [hunk 11]
```diff
@@ -433,4 +434,86 @@ export default class PartReviewService {
...
+
+  /**
+   * soft deletes an existing part review request if the requester, reviewer, or an admin initiates the request
+   * @param reviewRequestId - the ID of the part whose review request should be deleted
+   * @param user - the user attempting to delete the review request
+   * @param organizationId - the organization ID to validate permissions
+   * @returns the soft-deleted and transformed PartReviewRequest
+   */
+  static async deletePartReviewRequest(reviewRequestId: string, user: User, organizationId: string) {
+    const reviewRequest = await prisma.partReviewRequest.findFirst({
```
Comment (chpy04) [inline src/backend/src/services/part-review.services.ts]: Can be findUnique now instead of findFirst

File: src/backend/src/services/part-review.services.ts (hunk 7 above)
Comment (chpy04) [reply src/backend/src/services/part-review.services.ts:732]: Same thing on the specific DeletedException

File: src/backend/src/services/part-review.services.ts [hunk 12]
```diff
@@ -433,4 +434,86 @@ export default class PartReviewService {
...
+
+    if (!reviewRequest || reviewRequest.dateDeleted) {
+      throw new NotFoundException('Review request', reviewRequestId);
+    }
+
+    const isRequester = reviewRequest.requesterId === user.userId;
+    const isReviewer = reviewRequest.reviewerId === user.userId;
+    const isLeader =
+      (await userHasPermission(user.userId, organizationId, isLeadership)) ||
+      (await userHasPermission(user.userId, organizationId, isAdmin));
```
Comment (chpy04) [inline src/backend/src/services/part-review.services.ts]: and here, you don't need to check isAdmin

File: src/backend/src/services/part-review.services.ts [hunk 13]
```diff
@@ -433,4 +434,86 @@ export default class PartReviewService {
...
+      throw new NotFoundException('Part', partId);
+    }
+
+    const hasAccess =
+      (await userHasPermission(requester.userId, organizationId, isMember)) ||
+      (await userHasPermission(requester.userId, organizationId, isAdmin)) ||
+      (await userHasPermission(requester.userId, organizationId, isLeadership));
+
+    if (!hasAccess) {
+      throw new AccessDeniedAdminOnlyException('access part');
```
Comment (chpy04) [inline src/backend/src/services/part-review.services.ts]: Just throw an AccessDeniedException instead of an AccessDeniedAdminOnlyException, because more than just admins can do this

File: src/backend/src/services/part-review.services.ts [hunk 14]
```diff
@@ -433,4 +434,86 @@ export default class PartReviewService {
...
+    }
+
+    const isRequester = reviewRequest.requesterId === user.userId;
+    const isReviewer = reviewRequest.reviewerId === user.userId;
+    const isLeader =
+      (await userHasPermission(user.userId, organizationId, isLeadership)) ||
+      (await userHasPermission(user.userId, organizationId, isAdmin));
+
+    if (!isRequester && !isReviewer && !isLeader) {
+      throw new AccessDeniedAdminOnlyException('delete part review request');
```
Comment (chpy04) [inline src/backend/src/services/part-review.services.ts]: Same here, should be just AccessDeniedException

File: src/backend/src/transformers/part-review.transformer.ts [hunk 15]
```diff
@@ -67,7 +67,8 @@ export const partReviewRequestTransformer = (
     partId: reviewRequest.partId,
     requester: userTransformer(reviewRequest.requester),
     reviewerRequested: userTransformer(reviewRequest.reviewerRequested),
-    createdAt: reviewRequest.createdAt
+    createdAt: reviewRequest.createdAt,
+    dateDeleted: reviewRequest.dateDeleted
```
Comment (chpy04) [inline src/backend/src/transformers/part-review.transformer.ts]: Can you change this to reviewRequest.dateDeleted ?? undefined, we prefer undefined values to null values in the frontend

File: src/shared/src/types/part-review.types.ts [hunk 16]
```diff
@@ -44,6 +44,7 @@ export interface PartReviewRequest {
   requester: User;
   reviewerRequested: User;
   createdAt: Date;
+  dateDeleted?: Date | null;
```
Comment (chpy04) [inline src/shared/src/types/part-review.types.ts]: This is why we want the undefined in the transformer, then you don't need the | null here

**Review by chpy04 — CHANGES_REQUESTED** [review]
Few more tweaks

File: src/backend/src/services/part-review.services.ts (hunk 3 above)
Comment (chpy04) [reply src/backend/src/services/part-review.services.ts:683]: I should have been more specific here. If the part isn't in our database, we should throw a NotFoundException. If it is found but its deleted, we want to throw the DeletedException, so there should be two if statements

File: src/backend/src/services/part-review.services.ts (hunk 7 above)
Comment (chpy04) [reply src/backend/src/services/part-review.services.ts:732]: Same here

**Review by chpy04 — CHANGES_REQUESTED** [review]
One last minor thing

File: src/backend/src/services/part-review.services.ts (hunk 3 above)
Comment (chpy04) [reply src/backend/src/services/part-review.services.ts:683]: Other way around lol

**Review by chpy04 — CHANGES_REQUESTED** [review]
One last thing

**Review by chpy04 — APPROVED** [review]
Lgtm!

File: src/backend/src/controllers/part-review.controllers.ts [hunk 17]
```diff
@@ -159,4 +159,38 @@ export default class PartReviewController {
...
+    try {
+      const { reviewRequestId } = req.params;
+
+      const deleted = await PartReviewService.deletePartReviewRequest(
+        reviewRequestId,
+        req.currentUser,
+        req.organization.organizationId
+      );
+
+      res.status(200).json(deleted);
```
Comment (walker-sean) [inline src/backend/src/controllers/part-review.controllers.ts]: nit: Let's return a success message here instead

File: src/backend/src/prisma/migrations/migration_lock.toml
Comment (walker-sean) [inline src/backend/src/prisma/migrations/migration_lock.toml:1]: We need to leave this file

File: src/backend/src/routes/parts.routes.ts [hunk 18]
```diff
@@ -66,4 +66,13 @@ partsRouter.post(
 
 partsRouter.post('/common-mistake/:commonMistakeId/delete', PartReviewController.deleteCommonMistake);
 
+partsRouter.post(
+  '/reviewRequest/:partId/create',
```
Comment (walker-sean) [inline src/backend/src/routes/parts.routes.ts]: The `partId` should be first because its the identifier - so swap it with `reviewRequest`

File: src/backend/src/services/part-review.services.ts [hunk 19]
```diff
@@ -433,4 +448,90 @@ export default class PartReviewService {
...
+
+    if (part.dateDeleted) {
+      throw new DeletedException('Part', partId);
+    }
+
+    const role = await getUserRole(requester.userId, organizationId);
+    const hasAccess = isAtLeastRank(RoleEnum.MEMBER, role);
+
+    if (!hasAccess) {
+      throw new AccessDeniedException('access part');
```
Comment (walker-sean) [inline src/backend/src/services/part-review.services.ts]: Ok lol we've done this wrong in a bunch of places but the args that go into the `AccessDeniedException` constructor should be a full sentence

File: src/backend/src/services/part-review.services.ts [hunk 20]
```diff
@@ -433,4 +448,90 @@ export default class PartReviewService {
...
+    if (reviewRequest.dateDeleted) {
+      throw new DeletedException('Review request', reviewRequestId);
+    }
+
+    const isRequester = reviewRequest.requesterId === user.userId;
+    const isReviewer = reviewRequest.reviewerId === user.userId;
+    const isLeader = await userHasPermission(user.userId, organizationId, isLeadership);
+
+    if (!isRequester && !isReviewer && !isLeader) {
+      throw new AccessDeniedException('delete part review request');
```
Comment (walker-sean) [inline src/backend/src/services/part-review.services.ts]: Same here

**Review by walker-sean — CHANGES_REQUESTED** [review]

File: src/backend/src/controllers/part-review.controllers.ts (hunk 17 above)
Comment (nateci) [reply src/backend/src/controllers/part-review.controllers.ts]: Hey Sean, I went ahead and made this actually return a message? Not sure if that's what you meant, but if that's the case I can change for my other endpoint ticket as well.

File: src/backend/src/services/part-review.services.ts [hunk 21]
```diff
@@ -458,6 +475,92 @@ export default class PartReviewService {
...
+    if (reviewRequest.dateDeleted) {
+      throw new DeletedException('Review request', reviewRequestId);
+    }
+
+    const isRequester = reviewRequest.requesterId === user.userId;
+    const isReviewer = reviewRequest.reviewerId === user.userId;
+    const isLeader = await userHasPermission(user.userId, organizationId, isLeadership);
+
+    if (!isRequester && !isReviewer && !isLeader) {
+      throw new AccessDeniedException('Failed to delete part review request.');
```
Comment (walker-sean) [inline src/backend/src/services/part-review.services.ts]: nit: this is a permission error, not a failure - You should write this something like `only ... can do ....`

File: src/backend/src/services/part-review.services.ts [hunk 22]
```diff
@@ -458,6 +475,92 @@ export default class PartReviewService {
...
+
+    if (part.dateDeleted) {
+      throw new DeletedException('Part', partId);
+    }
+
+    const role = await getUserRole(requester.userId, organizationId);
+    const hasAccess = isAtLeastRank(RoleEnum.MEMBER, role);
+
+    if (!hasAccess) {
+      throw new AccessDeniedException('Failed to access part.');
```
Comment (walker-sean) [inline src/backend/src/services/part-review.services.ts]: This one can actually be a accessdeniedguestexception

**Review by walker-sean — CHANGES_REQUESTED** [review]

**Review by walker-sean — APPROVED** [review]
